pkg/reconciler/cache/publishedresources/replication/replication_controller.go

```diff
@@ -0,0 +1,224 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replication
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+
+	cacheclient "github.com/kcp-dev/kcp/pkg/cache/client"
+	"github.com/kcp-dev/kcp/pkg/cache/client/shard"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	cachev1alpha1 "github.com/kcp-dev/kcp/sdk/apis/cache/v1alpha1"
+	kcpclientset "github.com/kcp-dev/kcp/sdk/client/clientset/versioned/cluster"
+)
+
+const (
+	// ControllerName hold this controller name.
+	ControllerName = "kcp-published-resource-replication-controller"
+)
+
+// NewController returns a new replication controller.
+//
+// The replicated object will be placed under the same cluster as the original object.
+// In addition to that, all replicated objects will be placed under the shard taken from the shardName argument.
+// For example: shards/{shardName}/clusters/{clusterName}/apis/apis.kcp.io/v1alpha1/apiexports.
+func NewController(
+	shardName string,
+	dynamicCacheClient kcpdynamic.ClusterInterface,
+	kcpCacheClient kcpclientset.ClusterInterface,
+	gvr schema.GroupVersionResource,
+	replicated *ReplicatedGVR,
+	callback func(),
+) (*Controller, error) {
+	c := &Controller{
+		shardName: shardName,
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.DefaultTypedControllerRateLimiter[string](),
+			workqueue.TypedRateLimitingQueueConfig[string]{
+				Name: ControllerName,
+			},
+		),
+		dynamicCacheClient: dynamicCacheClient,
+		kcpCacheClient:     kcpCacheClient,
+		replicated:         replicated,
+		gvr:                gvr,
+		callback:           callback,
+	}
+
+	_, _ = c.replicated.Local.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueObject(obj, gvr) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueObject(obj, gvr) },
+		DeleteFunc: func(obj interface{}) { c.enqueueObject(obj, gvr) },
+	})
+
+	_, _ = c.replicated.Global.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueCacheObject(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueCacheObject(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueCacheObject(obj) },
+	})
```

Comment:
Will leak the handlers again when the controller is shutdown. 

Since the handlers are registered through members on the controller maybe register in `.Start`?
I'll put in a code suggestion